### PR TITLE
Access control

### PR DESCRIPTION
**Pull request by MGathier**
Added context to user roles. 
Updated mapping from path to role using new tables.
Also for #366: More fine-grained access control for applications

### CODE REVIEW

File: axonserver/src/main/java/io/axoniq/axonserver/AxonServerAccessController.java [hunk 1]
```diff
@@ -25,9 +26,27 @@
...
     boolean allowed(String fullMethodName, String context, String token);
 
-    Collection<PathMapping> getPathMappings();
-
     boolean isRoleBasedAuthentication();
 
-    Set<String> getRoles(String token, String context);
+    Set<String> getRoles(String token);
+
+    default Set<String> rolesForOperation(String permission) {
```
Comment (gjmwoods) [inline axonserver/src/main/java/io/axoniq/axonserver/AxonServerAccessController.java]: nit: this class could do with some javadoc

File: axonserver/src/main/java/io/axoniq/axonserver/access/user/UserController.java [hunk 2]
```diff
@@ -58,14 +61,32 @@ private User syncUser(String username, String password, String[] roles) {
     }
 
     @Transactional
-    public User updateUser(String username, String password, String[] roles) {
```
Comment (gjmwoods) [inline axonserver/src/main/java/io/axoniq/axonserver/access/user/UserController.java]: nit: javadoc

File: axonserver/src/main/java/io/axoniq/axonserver/rest/RoleRestController.java [hunk 3]
```diff
@@ -38,15 +38,15 @@ public RoleRestController(RoleController roleController) {
     @GetMapping("user")
     Collection<String> listUserRoles(){
```
Comment (gjmwoods) [inline axonserver/src/main/java/io/axoniq/axonserver/rest/RoleRestController.java]: nit: javadoc

Comment (gjmwoods) [reply axonserver/src/main/java/io/axoniq/axonserver/rest/RoleRestController.java]: public?

File: axonserver/src/test/java/io/axoniq/axonserver/access/role/RoleControllerTest.java [hunk 4]
```diff
@@ -41,29 +33,29 @@
 
     private RoleController roleController;
 
-    @Autowired
-    private RoleRepository roleRepository;
+//    @Autowired
```
Comment (gjmwoods) [inline axonserver/src/test/java/io/axoniq/axonserver/access/role/RoleControllerTest.java]: test ought to be removed if no longer needed or replaced

File: axonserver/src/main/java/io/axoniq/axonserver/access/jpa/UserRole.java [hunk 5]
```diff
@@ -32,19 +32,27 @@
...
+    public UserRole(User user, String role, String context) {
         this.user = user;
         this.role = role;
+        this.context = context;
     }
 
     public UserRole() {
     }
 
+    public UserRole(String context, String role) {
```
Comment (saratry) [inline axonserver/src/main/java/io/axoniq/axonserver/access/jpa/UserRole.java]: is this a value object i.o. an entity?
the bi-directional reference to User is not really needed IMO.

File: axonserver/src/main/java/io/axoniq/axonserver/access/jpa/User.java [hunk 6]
```diff
@@ -108,4 +105,25 @@ public boolean equals(Object o) {
...
+
+        return newUser;
+    }
+
+    public void removeContext(String context) {
+        for (Iterator<UserRole> contextIterator = roles.iterator(); contextIterator.hasNext(); ) {
+            UserRole userRole = contextIterator.next();
+            if (userRole.getContext().equals(context)) {
+                contextIterator.remove();
+                userRole.setUser(null);
```
Comment (saratry) [inline axonserver/src/main/java/io/axoniq/axonserver/access/jpa/User.java]: if the UserRole was a value object, we could simply remove it from the collection.

File: axonserver/src/main/java/io/axoniq/axonserver/access/jpa/User.java [hunk 7]
```diff
@@ -108,4 +105,25 @@ public boolean equals(Object o) {
     public int hashCode() {
         return Objects.hash(userName);
     }
+
+    public static User newContextPermissions(User user) {
```
Comment (saratry) [inline axonserver/src/main/java/io/axoniq/axonserver/access/jpa/User.java]: Maybe useful to add JavaDocs here.

File: axonserver/src/main/java/io/axoniq/axonserver/access/roles/FunctionRole.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.Id;
+import javax.persistence.JoinColumn;
+import javax.persistence.ManyToOne;
+import javax.persistence.Table;
+
+/**
+ * Mapping between a function name and a access role. Mapping is many-to-many.
```
Comment (saratry) [inline axonserver/src/main/java/io/axoniq/axonserver/access/roles/FunctionRole.java]: It is not mapped as many to many

File: axonserver/src/main/java/io/axoniq/axonserver/access/roles/Role.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+ * @since 4.2
+ */
+@Entity
+@Table(name = "ROLES")
+public class Role {
+
+    @Id
+    private String role;
+
+    private String admin;
```
[the remark this answers is not in the document]
Comment (MGathier) [reply axonserver/src/main/java/io/axoniq/axonserver/access/roles/Role.java]: Deleted the field.

File: axonserver/src/main/java/io/axoniq/axonserver/access/roles/Role.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (c) 2017-2019 AxonIQ B.V. and/or licensed to AxonIQ B.V.
+ * under one or more contributor license agreements.
+ *
+ *  Licensed under the AxonIQ Open Source License Agreement v1.0;
+ *  you may not use this file except in compliance with the license.
+ *
+ */
+
+package io.axoniq.axonserver.access.roles;
```
[the remark this answers is not in the document]
Comment (MGathier) [reply axonserver/src/main/java/io/axoniq/axonserver/access/roles/Role.java]: moved entities to JPA package.

File: axonserver/src/main/java/io/axoniq/axonserver/access/jpa/UserRole.java (hunk 5 above)
Comment (MGathier) [reply axonserver/src/main/java/io/axoniq/axonserver/access/jpa/UserRole.java]: Deleted bi-di reference

File: axonserver/src/main/java/io/axoniq/axonserver/access/jpa/User.java (hunk 6 above)
Comment (MGathier) [reply axonserver/src/main/java/io/axoniq/axonserver/access/jpa/User.java]: done

File: axonserver/src/main/java/io/axoniq/axonserver/access/roles/FunctionRole.java (hunk 8 above)
Comment (MGathier) [reply axonserver/src/main/java/io/axoniq/axonserver/access/roles/FunctionRole.java]: updated comment.

**Pull request merged**